In [7]:
import pandas as pd
import numpy as np
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
import pandas as pd
import numpy as np
from sklearn.preprocessing import scale,minmax_scale,robust_scale
#%matplotlib inline background 그래프 실행 및 결과 확인
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.formula.api import ols
from statsmodels.api import qqplot, add_constant
import statsmodels.formula.api as smf
from statsmodels.tools.eval_measures import rmse
#예측
from sklearn.tree import export_graphviz 

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor as RFR
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import GridSearchCV
#평가함수
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import accuracy_score
#모델 성능 평가
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
#confusion matrix
from sklearn.metrics import confusion_matrix

#Scale 변환:평균,표준편차 기준
from sklearn.preprocessing import StandardScaler
#주성분 분석
from sklearn.decomposition import PCA
#회귀분석
import statsmodels.api as sm
#RMSE:root mean squared error 함수
def root_mean_squared_error(y_true,y_pred):
    return np.sqrt(mean_squared_error(y_true,y_pred))
#MAPE:mean absolute percentage error 함수
def mean_absolute_percentage_error(y_true,y_pred):
    return np.mean(np.abs((y_true-y_pred)/y_true))*100
#분류 : KNN
from sklearn.neighbors import KNeighborsClassifier
#로지스틱 회귀
from statsmodels.api import Logit
#데이터 분항
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import confusion_matrix
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from scipy.cluster.hierarchy import linkage, dendrogram, fcluster
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans, DBSCAN
import pandas as pd

import matplotlib.pyplot as plt
import matplotlib

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

# 데이터 분할
from sklearn.model_selection import train_test_split


from sklearn.metrics import f1_score, roc_curve, auc, precision_score, recall_score
import matplotlib.pyplot as plt
import os

# confusion matrix
from sklearn.metrics import confusion_matrix

import numpy as np
import warnings

warnings.filterwarnings("ignore")

from mlxtend.preprocessing import TransactionEncoder




In [ ]:
df=pd.read_csv("/home/piai/Desktop/POSCO_Ai-BigData_academy/Week2_BigData/실습파일/HMEQ.csv")

In [5]:
import os
matplotlib.rc("font", family="NanumGothic")
plt.rcParams['axes.unicode_minus']=False
os.environ['PATH']+=os.pathsep+"~/anaconda3/lib/python3.7/site-packages"

##  연관규칙

In [8]:
df_raw=[['우유','라면'],
       ['라면'],
       ['아이스크림','우유'],
       ['과자','라면'],
       ['아이스크림']]
df_raw


[['우유', '라면'], ['라면'], ['아이스크림', '우유'], ['과자', '라면'], ['아이스크림']]

In [9]:

#Item 중 index 0번을 갖고 있으면, True, 아니면 False
enc = TransactionEncoder()
df_raw_enc = enc.fit_transform(X=df_raw)


df_asso = pd.DataFrame(df_raw_enc, columns=enc.columns_)
df_asso.head()

과자     라면  아이스크림     우유
0  False   True  False   True
1  False   True  False  False
2  False  False   True   True
3   True   True  False  False
4  False  False   True  False

In [11]:
# 최소 지지도
min_support = 0.1

# 지지도 계산
df_freq = apriori(df_asso, min_support = min_support, use_colnames=True)
df_freq.round(3)

support     itemsets
0      0.2         (과자)
1      0.6         (라면)
2      0.4      (아이스크림)
3      0.4         (우유)
4      0.2     (과자, 라면)
5      0.2     (우유, 라면)
6      0.2  (아이스크림, 우유)

In [24]:
# 기준 = 신뢰도, 최소 신뢰도(threshold) =0.5
df_asso_rule = association_rules(df_freq, metric="confidence",min_threshold=0.5)
df_asso_rule.round(3).sort_values(by='confidence', ascending=False)

antecedents consequents  antecedent support  consequent support  support  \
0        (맥주)        (콜라)               0.500               0.667      0.5   
1        (콜라)        (맥주)               0.667               0.500      0.5   
2        (소주)        (콜라)               0.667               0.667      0.5   
3        (콜라)        (소주)               0.667               0.667      0.5   

   confidence   lift  leverage  conviction  
0        1.00  1.500     0.167         inf  
1        0.75  1.500     0.167       2.000  
2        0.75  1.125     0.056       1.333  
3        0.75  1.125     0.056       1.333

지지도 신뢰도 향상도
맥주 콜라 = 신뢰도1.0 맥주를 구매한 고객은 콜라도 반드시 구매 =>맥주구매한 고객에게 콜라추천
콜라 맥주 = 신뢰도 0.75 =콜라를 구매한 고객 중 맥주를 고른 고객 0.75 =lift(1.5)콜라구매고객에게 소주보다 맥주 추천
콜라 소주 = Lift 1.125 

## 모델 평가 

In [25]:
print(f1_score(df_test_y, y_pred))

NameError: name 'df_test_y' is not defined

In [51]:
y=pd.Series([1,2,3,4])

In [52]:
y_pred=np.array([4,2,3,4])

In [50]:
type(y_pred)

numpy.ndarray

In [53]:
fpr, tpr, thresholds =roc_curve(y, y_pred)

ValueError: multiclass format is not supported

In [39]:
y

0    1
1    2
2    3
3    4
dtype: int64

In [40]:
y_pred

(4, 2, 3, 4)

In [28]:
print(f1_score(y,y_pred))

ValueError: Target is multiclass but average='binary'. Please choose another average setting, one of [None, 'micro', 'macro', 'weighted'].

### 군집알고리즘

In [55]:
df_raw=pd.read_csv("/home/piai/Desktop/POSCO_Ai-BigData_academy/Week2_BigData/실습파일/직무능력평가.csv")

In [56]:
df_raw_x = df_raw.drop('EVAL', axis=1)
df_raw_y = df_raw['EVAL']
df_raw_y.head()

0    43
1    63
2    71
3    61
4    81
Name: EVAL, dtype: int64

In [57]:
df_raw_x.corr().round(3)

DISSATIS  PRIVILEGE  CHANCE  ADVANCE  CRITISM  PROMOTION
DISSATIS      1.000      0.558   0.597    0.669    0.188      0.225
PRIVILEGE     0.558      1.000   0.493    0.445    0.147      0.343
CHANCE        0.597      0.493   1.000    0.640    0.116      0.532
ADVANCE       0.669      0.445   0.640    1.000    0.377      0.574
CRITISM       0.188      0.147   0.116    0.377    1.000      0.283
PROMOTION     0.225      0.343   0.532    0.574    0.283      1.000

In [58]:
x_cols = df_raw_x.columns
scaler = StandardScaler()
nd_scaled = scaler.fit_transform(df_raw_x)
print('Scaled Data:\n{}'.format(nd_scaled[:].round(3)))
print('Shape : \n{}'.format(nd_scaled.shape))

Scaled Data:
[[-1.192 -1.923 -1.505 -0.355  1.771  0.204]
 [-0.199 -0.177 -0.205 -0.16  -0.182  0.402]
 [ 0.26   1.236  1.095  1.112  1.155  0.501]
 [-0.275 -0.676 -0.812 -1.04   0.949 -0.784]
 [ 0.871  0.238  0.835  0.623  0.846  0.402]
 [-0.886 -0.344 -1.072 -1.04  -2.649 -0.883]
 [ 0.031 -0.925 -0.032  0.134 -0.696 -0.784]
 [ 0.642 -0.26  -0.118  0.525 -0.901 -0.191]
 [ 1.176  1.568  0.921  0.623  0.846 -1.18 ]
 [-0.428 -0.676 -0.812 -0.258  0.538 -0.191]
 [-1.039 -0.011  0.142 -0.649 -0.798 -0.883]
 [-0.504 -0.51  -1.505 -0.551 -0.079 -0.191]
 [-0.351  0.321 -1.245 -0.942 -1.209 -1.773]
 [ 1.253  2.483 -0.985 -0.551  0.23  -0.784]
 [ 0.794  0.072  1.355  1.405  0.23   0.303]
 [ 1.787 -0.26   1.355 -0.453 -2.135 -0.685]
 [ 1.406  0.903  1.095  1.405  0.435  1.984]
 [-0.504  0.986  1.615 -0.942  0.538  1.687]
 [ 0.26  -0.593  0.055  1.014  1.052  0.303]
 [-0.657  1.236 -0.205 -0.062  0.332  0.896]
 [-2.032 -1.674 -1.938 -2.116 -1.107 -0.982]
 [-0.428 -0.094  0.488  0.134  0.538 -0.19